### ISSUE TITLE
Tracking issue for eRFC 2497, "if- and while-let-chains, take 2"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#53667]
This is a tracking issue for the eRFC *"if- and while-let-chains, take 2"* (rust-lang/rfcs#2497).
For the tracking issue for the immediate edition changes, see #53668.

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on forge][doc-guide]) 
- [ ] Formatting for new syntax has been added to the [Style Guide] ([nightly-style-procedure])
- [ ] resolve expr fragment specifier issue ([#86730](https://github.com/rust-lang/rust/issues/86730))
- [x] Initial stabilization PR: #94927, reverted in #100538
- [ ] Second stabilization PR: https://github.com/rust-lang/rust/pull/132833

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation
[nightly-style-procedure]: https://github.com/rust-lang/style-team/blob/master/nightly-style-procedure.md 
[Style Guide]: https://github.com/rust-lang/rust/tree/master/src/doc/style-guide

**Unresolved questions:**

- [x] [The final syntax](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#the-final-syntax)
    - This would benefit from a design note summarizing previous discussions on `let` as a boolean expression, and possibilities of `is` syntax.
- [x] [Should temporary and irrefutable lets without patterns be allowed in some form?](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#irrefutable-let-bindings-after-the-first-refutable-binding)
- [x] [Chained `if let`s inside `match` arms](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#chained-if-lets-inside-match-arms)

**Collected issues:**

- [x] https://github.com/rust-lang/rust/issues/60336
- [x] https://github.com/rust-lang/rust/issues/60707
- [x] https://github.com/rust-lang/rust/issues/60254
- [x] https://github.com/rust-lang/rust/issues/60571
- [x] https://github.com/rust-lang/rust/issues/44990

**Implementation history:**

- https://github.com/rust-lang/rust/pull/59290
- https://github.com/rust-lang/rust/pull/59439
- https://github.com/rust-lang/rust/pull/60225
- https://github.com/rust-lang/rust/pull/59288
- https://github.com/rust-lang/rust/pull/60861
- https://github.com/rust-lang/rust/pull/61988
- https://github.com/rust-lang/rust/pull/79328
- https://github.com/rust-lang/rust/pull/80357
- https://github.com/rust-lang/rust/pull/82308
- https://github.com/rust-lang/rust/pull/88642
- https://github.com/rust-lang/rust/pull/93086
- https://github.com/rust-lang/rust/pull/93213
- https://github.com/rust-lang/rust/pull/97295
- https://github.com/rust-lang/rust/pull/94754
- https://github.com/rust-lang/rust/pull/94974
- https://github.com/rust-lang/rust/pull/95008
- https://github.com/rust-lang/rust/pull/95314

**Unresolved problems**

- [ ] Can we be confident that the implementation is correct and well tested?

**Comment by Centril** [comment]
I'm working on implementing this in a 3+ PR step fashion based on discussions with @oli-obk.

**Comment by oli-obk** [comment]
for the record, the discussed step list is

1. remove `if` from `HIR` and make lowering emit a `match`
2. remove `if let` from `AST` and instead have a `let` expression
3. "do the rest of the work" (might get split up further).

**Comment by alexreg** [comment]
I can't imagine 3 would be a particularly large step (enough to merit splitting), but 3 steps sounds fair enough.

**Comment by estk** [comment]
@Centril I'm really jazzed about this feature, lmk if there is any work I can help with.

**Comment by Centril** [comment]
@estk :) The work is pretty non-parallelizable at the moment; However, once https://github.com/rust-lang/rust/pull/60861 lands you may want to take a look at https://github.com/rust-lang/rust/issues/60336.

**Comment by Manishearth** [comment]
Question worth discussing: This introduces `let` as something _like_ an expression in the AST.

Do we plan to make it a true expression? As in, will `$e:expr` match `let Pat(..) = foo` in macros? This does not mean that it will _work_ everywhere (as @centril correctly points out, `$t:ty` matches `impl Trait`, but it's only allowed in certain positions in the AST), it just means that macros will be able to match on it. They may error further down the pipeline if it gets used outside of an `if`.

This is technically a breaking behavior change; a macro with a fallthrough to `let $p:pat = $e:expr` will suddenly start behaving differently.

**Comment by eddyb** [comment]
@Manishearth I *thought* that you can't probe whether something parses as an AST fragment in `macro_rules` macros, only tokens (this includes `$i:ident`, `$l:lit`, etc.).
That is, I expected the first `expr`/`pat`/`stmt`/`block`/`item`/etc. that is reached to be final.
But I'm wrong, and this is either behavior introduced in the past couple years, or I was just confused.

**Comment by petrochenkov** [comment]
>this is either behavior introduced in the past couple years

This is https://github.com/rust-lang/rust/pull/42913 if I'm not mistaken, almost 2 years ago.
One token lookahead is used to reject alternatives that "obviously don't fit".

**Comment by matklad** [comment]
(not sure this is the right place for this, but I don't think we have anything else opened for "pattern matching expressions" discussion).

I've just noticed that Java is getting [pattern matching](https://openjdk.java.net/jeps/305) along the lines proposed in https://github.com/rust-lang/rfcs/issues/929#issuecomment-285602496. That is, control-flow sensitive scoping and patterns on the right are now almost Java-level proven and trusted language design choices.

**Comment by dzmitry-lahoda** [comment]
any actual syntax for this in nightly rust? just cannot type it to work.

**Comment by mackycheese21** [comment]
Any progress updates? Although I love rust this is by far the jankiest part  of the language

**Comment by c410-f3r** [comment]
I am trying to tackle this feature one bit at a time in my spare time. No promises for a concrete release date though

**Comment by HTG-YT** [comment]
Currently if I had to match against two different `Option<T>`s I'd have to:

```rs
if let (Some(smth), Some(another_smth)) = (smth_option(), another_smth_option()) {
    /* more code */
}
```

IMHO this works, but not an ideal way to me. I like the idea of this RFC. Looking forward to this in Rust! 👍

**Comment by overlisted** [comment]
@HTG-YT But you can't do something like this:
```rust
if let (Some(smth), Some(another_smth)) = (smth_option(), another_smth_option(smth)) {
    /* more code */
}
```

It's what I will be using this feature for the most.

**Comment by c410-f3r** [comment]
Just to make it official

@rustbot claim

**Comment by joshtriplett** [comment]
@c410-f3r How goes the development on this? (No rush, just checking in.)

**Comment by hddharvey** [comment]
It was not clear to me reading the RFC (https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md) whether the proposed precedence rules (`=` having higher precedence than `&&`) applies to regular `let` statements too, or just in an `if` or `while` context.

E.g., will

```
let boolean_value = returns_boolean() && also_returns_boolean();
```

be affected and now require parenthesis to disambiguate?

**Comment by teohhanhui** [comment]
> To be precise, the changes in this RFC entail that || has the lowest precedence at the top level of if STUFF { .. }. The operator && has then the 2nd lowest precedence and binds more tightly than ||. If the user wants to disambiguate, they can write (EXPR && EXPR) or { EXPR && EXPR } explicitly. The same applies to while expressions.

It does mention `if` and `while` explicitly.

**Comment by foodornt** [comment]
it's been 3 years, what happened to this RFC 😢

**Comment by HKalbasi** [comment]
I wrote a [RFC draft about let expression](https://github.com/HKalbasi/rfcs/blob/master/text/0000-let-expression.md). I like to hear comments of people involved here before submitting it officially.

**Comment by HKalbasi** [comment]
Now I submitted it [in this PR](https://github.com/rust-lang/rfcs/pull/3159). Thanks for your work on `if let chain` and I hope this RFC is a positive step in this direction and for the rust in general.

**Comment by camsteffen** [comment]
I opened rust-dev-tools/fmt-rfcs#169 to discuss a formatting question.

**Comment by plazmoid** [comment]
Why `||` is forbidden in let chains? I didn't find anything about it in RFC

**Comment by c410-f3r** [comment]
> Why `||` is forbidden in let chains? I didn't find anything about it in RFC

https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#keeping-the-door-open-for-if-let-or-expressions

"Not supported" is probably a better term

**Comment by c410-f3r** [comment]
A stabilization attempt is available at https://github.com/rust-lang/rust/pull/94927

**Comment by zertyz** [comment]
The compiler lead me here after a condition which, intuitively I wrote like this **(a)**:
```
if let Some(x) = x && let Some(y) = f(x) {
    (1)
} else {
    (2)
}
```

It seems the only alternative, by now, is to write it like this **(b)**:
```
if let Some(x) = x {
    if let Some(y) = f(x) {
        (1)
    } else {
        (2)
    }
} else {
    (2)
}
```

... so I believe this 2018 "take 2" is not trying to address the issue by interpreting usages like **(a)** as a syntactic sugar for the true valid expression **(b)**... ?

Maybe it is about time for a "take 3" ?

**Comment by WaffleLapkin** [comment]
@zertyz there is a stabilization attempt @ https://github.com/rust-lang/rust/pull/94927 that targets Rust 1.62 (see the comment just before yours), so I wouldn't call this "stalled".

**Comment by januwA** [comment]
Does this expression work in rust 1.62

```rs
if let Some(email) = data.user.email && !email.is_empty() {
    q = q.filter(users::email.eq(email));
}
```

I now get an error in 1.60

```
error[E0658]: `let` expressions in this position are unstable
```

**Comment by c410-f3r** [comment]
2 years, 4 months, 3 weeks and 1 day of long nights, obstacles and headaches.

Hope stabilization won't be reverted but regardless, thanks to everyone who helped make this feature a reality.

**Comment by mousetail** [comment]
> ~2 years, 4 months, 3 weeks and 1 day of long nights, obstacles and headaches.~
> 
> ~Hope stabilization won't be reverted but regardless, thanks to everyone who helped make this feature a reality.~
> 
> EDIT: Reverted in #100538

Celebrated too soon :(

**Comment by Mark-Simulacrum** [comment]
Reopening this tracking issue, since it's not currently stable.

I'd also like to flag https://github.com/rust-lang/rust/pull/94927#issuecomment-1165156328 which leads me to believe that perhaps stabilization was actually premature for that reason as well? It's not clear to me that we ever had a follow-up to #97295 which actually adjusted our restrictions further, and [this error](https://github.com/rust-lang/rust/blob/aa0b40ef6c70f29a2c98a9faa2b6b7cd706ab120/src/test/ui/mir/issue-92893.stderr#L15-L22) leads me to believe that let expressions would have been stable in more places than I expected us to *want*. @c410-f3r -- can you elaborate on whether there was a follow-up to #97295 that resolved @compiler-errors concern?

Maybe that was https://github.com/rust-lang/rust/pull/98633, but given the error I mentioned in that random MIR test it seems like it may have been (still) incomplete?

**Comment by c410-f3r** [comment]
#98633 was merged after #97295 and then the underlying solution was refined in #99731 and #100011.

`issue-92893.stderr` now correctly denies undesired `let` expressions in the parser as well as other `let` locations like `if let Some(elem) = _opt && [1, 2, 3][let _ = ()] = 1`. Therefore, it seems that everything related to https://github.com/rust-lang/rust/pull/94927#issuecomment-1099605024 is fixed.

@Mark-Simulacrum Can you close https://github.com/rust-lang/rust/issues/60707 and https://github.com/rust-lang/rust/issues/60571? These issues are no longer relevant.

**Comment by Mark-Simulacrum** [comment]
Ah, I missed that we're emitting a "spurious" feature gate error even though the syntax itself is denied in a separate error in that case. Seems OK.

**Comment by camsteffen** [comment]
I just tried writing `if <BOOL> && if let ...` and it fails pretty terribly. Shouldn't this be supported? If not, there should at least be better diagnostics.

```rust
fn foo(b: bool, x: Option<u32>) {
    if b && if let Some(x) = x {}
}
```

```text
error: expected `{`, found `}`
 --> src/lib.rs:3:1
  |
3 | }
  | ^ expected `{`
  |
note: the `if` expression is missing a block after this condition
 --> src/lib.rs:2:8
  |
2 |     if b && if let Some(x) = x {}
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by tronta** [comment]
maybe you should try
``` rust
fn foo(b: bool, x: Option<u32>) {
    if b && let Some(x) = x {}
}
```

**Comment by camsteffen** [comment]
Oh no.

**Comment by joshtriplett** [comment]
@camsteffen I think it'd be worth improving the diagnostics for that case; perhaps we could detect it and give a rustfix-able suggestion for "delete the `if`"?

**Comment by camsteffen** [comment]
Yes I was thinking that. Seems tricky though since the parser will be in a nested if expression.

**Comment by alercah** [comment]
This isn't specific to `if let`: `if true && if true {false}` gives the same error.

It would probably have been nice to disallow this long ago by requiring parens for any nested block expression, as is done for struct expressions, but that ship is long sailed. Definitely should be considered for improved diagnostics and maybe a lint, but it shouldn't affect this RFC.

**Comment by lunabunn** [comment]
What is the status on this? Is it ready for a FCP?

**Comment by rami3l** [comment]
@lunabunn 
According to https://github.com/rust-lang/rust-clippy/issues/9354, the feature was once stabilized, and reverted due to the issue https://github.com/rust-lang/rust/issues/100513. However the latter is closed now. So I guess we should be back on the right track now.

**Comment by Lee-Janggun** [comment]
There were a few ICEs related to this feature, including the following I could find.
- #100513
- #99852
- #100276
- #99938

After the remaining open issues are closed, another FCP may be possible (hopefully ready for 1.65).

**Comment by est31** [comment]
Now with #103034 merged, only #99938 is open of your list @Lee-Janggun , which should be fairly easy to close.

I've also filed:

* #103248

It should ideally be resolved before stabilization. Also, while #100513 is closed, it should probably be reopened because the drop order isn't perfect yet.

**Comment by solomatov** [comment]
What are the plans to stabilize it. Is there any chance it gets into 1.66?

**Comment by jhpratt** [comment]
Considering 1.66 beta branched on Friday, there is zero chance it'll land then.

**Comment by est31** [comment]
As @jhpratt pointed out, it won't stabilize for 1.66 now.

The implementation state has recently been greatly improved by #103034 and #102998. There has also been lang team feedback on the open questions about the drop order.

There is one weird quirk remaining, which is around bool expressions. My PR #103293 tries to address it. The quirk also occurs in non-let if chains, it would still be great to have it fixed before let chains stabilize as it makes the bool conditionals interact with the `let` conditionals, which they otherwise don't do: the bool conditionals drop in a totally different pattern from the `let` conditionals, and at a totally different time. This separation makes sense as bool conditionals don't leave behind any bindings while let conditionals do. But with the && twist present, adding a `let _ = () &&` conditional in front of a bunch of bool conditionals changes the drop order within the group of bool conditionals. That seems uglier to me than just plain 

There is still one big-ish bug in #103476. We might settle on it being an acceptable quirk (like #103784 for example), but ideally it would be investigated further before let chains stabilize.

Outside of that I don't know of any blockers.

**Comment by est31** [comment]
As a small new years update on the let-chains feature. Some new developments:

* An [ICE](https://github.com/rust-lang/rust/issues/104843) was discovered. @dingxiangfei2009 has assigned themselves. According to them, it is reminiscent of an [ICE with the let else feature](https://github.com/rust-lang/rust/issues/102317) which they have fixed, so they know how to approach the issue.
* In general, it would be good to have people *test* let chains, to discover issues like the one above while the feature is still unstable. If you target nightly, then you can use let chains directly. If you target stable, you can add if chains to your code via the `if_chain` crate. Then, if it's open source, you can just put a comment into the [dedicated issue](https://github.com/rust-lang/rust/issues/104893), or alternatively you can run the [if-to-let-chain](https://github.com/Alexendoo/if-to-let-chain) tool on your code base yourself, and report any ICEs or lifetime bugs you encounter. I've for example done a [migration of clippy](https://github.com/rust-lang/rust/issues/104893#issuecomment-1336518106) with a 9k line diff to find more ICEs like the one above, but I couldn't find any.
* My [PR](https://github.com/rust-lang/rust/pull/103293) to change the drop order of && and || was merged and is now riding the trains, currently on the beta branch. The change affects both if-chain and normal && and || usage, but consistency is especially important for if-chains as their drop order was interacting with the twist in a complicated manner: this is now simpler with the change.
* There has been only minor progress on the issue https://github.com/rust-lang/rust/issues/103476 .

**Comment by Boiethios** [comment]
I am not sure it has been discussed before, and I know it's a syntax which shouldn't be used, but isn't that a bit confusing?

```rust
let condition = false;
let some_other_condition = false;

if let false = condition && some_other_condition {
    println!("step 1"); // Does not enter here
}

if let false = (condition && some_other_condition) {
    println!("step 2"); // Enters here
}
```

The first notation is ambiguous, and maybe not everyone expect the precedence to go that way. At least, there should be some kind of warning, like “You better write it that way: `if !condition && some_other_condition { … }` if that's what you meant”. That warning may even be shown whenever one pattern matches a `bool`, regardless of the context.

**Comment by kamulos** [comment]
Wouldn't that be the same as the following:

```rust
    let condition = false;
    let some_other_condition = false;

    let a = condition && some_other_condition == false;
    let b = (condition && some_other_condition) == false;

    println!("{a}, {b}");
```

**Comment by jtran** [comment]
Interesting.  The above comments are saying that the `=` operator in an `if let pat = expr` can be thought of as similar to `==` in terms of precedence level, not like `=` in assignment statements.  That is surprising to me.

I suspect that it was designed this way since it's very common to write code that looks like:

```rust
if let Some(x) = foo() && bar(x) {
```

If the precedence were the other way, the above code would require parentheses.

The relevant section of the RFC is [here](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#dealing-with-ambiguity).  It also came up in this thread https://github.com/rust-lang/rust/issues/53667#issuecomment-860336898.

**Comment by rsalmei** [comment]
> ```rust
> if let Some(x) = foo() && bar(x) {
> ```

If the precedence were the other way, it wouldn't be possible to bind `x` and use it on the next expression(s).

**Comment by kristof-mattei** [comment]
> I am not sure it has been discussed before, and I know it's a syntax which shouldn't be used, but isn't that a bit confusing?
> 
> ```rust
> let condition = false;
> let some_other_condition = false;
> 
> if let false = condition && some_other_condition {
>     println!("step 1"); // Does not enter here
> }
> 
> if let false = (condition && some_other_condition) {
>     println!("step 2"); // Enters here
> }
> ```
> 
> The first notation is ambiguous, and maybe not everyone expect the precedence to go that way. At least, there should be some kind of warning, like “You better write it that way: `if !condition && some_other_condition { … }` if that's what you meant”. That warning may even be shown whenever one pattern matches a `bool`, regardless of the context.

Correct me if I'm wrong but the way I read the `let` statement is as follows: does the pattern match? `let false = false` matches. Just like `let None = None`. So now we have `true && some_other_condition`, which is `false` so we don't enter the `if` statement.

For the second one we have `(false && false)`, which is `false` which matches the `let false = false` predicate.

**Comment by mark-i-m** [comment]
I think people could reasonably interpret it either way, but it probably auto to be a warning or at least a clippy lint.

**Comment by RealRTTV** [comment]
Given that without let chains, the only syntactically valid code snippet is the second one, that to not cause breaking changes the only way to not cause breaking let chains is to make the first example the correct one?

**Comment by PoignardAzur** [comment]
Any progress on this feature?

**Comment by rami3l** [comment]
> Any progress on this feature?

@PoignardAzur AFAIK this feature is still blocked on https://github.com/rust-lang/rust/issues/103476.

A simple ping in this case does nothing. If you'd like to see this feature stabilized, I recommend you to:

> try out let chains in your own code, and report bugs you find.

_Originally posted by @est31 in https://github.com/rust-lang/rust/issues/103476#issuecomment-1338322481_

**Comment by est31** [comment]
The main/only blocker at this point that I see is #104843, for which there is #107251, which is blocked on a nice overview comment as asked for in https://github.com/rust-lang/rust/pull/107251#issuecomment-1779644448, then an FCP.

At this point, I don't think that #103476 is a blocker any more, as enough time has passed. For #104893, I think due to the passage of time more people have tried out let chains, at least [grep.app](https://grep.app/search?q=feature%28let_chains%29&filter[lang][0]=Rust) shows plenty of users. I don't really think it's much of a blocker either at this point any more.

**Comment by pnkfelix** [comment]
re todo item "resolve expr fragment specifier issue", see also https://github.com/rust-lang/rfcs/pull/3531, "Macro matcher fragment specifiers edition policy".

**Comment by photino** [comment]
This feature is really helpful. Is it possible to use in the Edition 2024?

**Comment by slanterns** [comment]
It's not related to edtion.

**Comment by inquisitivecrystal** [comment]
> This feature is really helpful. Is it possible to use in the Edition 2024?

Hi!

No, I'm afraid not. Editions have to do with changes that might break old code; as far as I know, this is completely unrelated. Also, people generally prefer that this sort of question not be asked on the tracking issue, as it sends a notification out to everyone subscribed to the thread without providing new information. It's no big deal, just something to be careful about going forward.

I believe the current status is described in est31's last post above.

**Comment by sammysheep** [comment]
Nightly user here.  My experience with this feature has been very good. I'm surprised with how concisely I can convey my intent.

I find I tend to use chains with 3 clauses for some reason, and the succinct transitive nature of it feels elegant.

10/10. Would recommend!

**Comment by Nilstrieb** [comment]
@joshtriplett has opened https://github.com/rust-lang/rfcs/pull/3573 and it is not clear whether it makes sense to have both features (I personally don't think so).

**Comment by est31** [comment]
Yeah before we stabilize this we should in any case wait for the decision on that RFC, which in this instance can turn in any direction.

**Comment by roylaurie** [comment]
> Yeah before we stabilize this we should in any case wait for the decision on that RFC, which in this instance can turn in any direction.

I think this should keep moving forward. Why trash all of this work unless the other RFC actually gets accepted?

The other idea has already been shot down before. The recent sentiment I've seen on Reddit and in the comments on Github don't paint a picture of anything changing in that regard.

**Comment by jpmckinney** [comment]
I also say keep moving forward. In terms of governance, I don’t think it makes sense to pause work on one RFC anytime another RFC suggests an alternative. That’ll just stall everything. (Also, that other RFC will not pause for this one.)

**Comment by TennyZhuang** [comment]
I do not think rust-lang/rfcs#3573 should block this RFC. `is` operator is an alternative to the existing feature `if let`, instead of `let-chains`. The motivation of the new `is` operator is better readability for a left-to-right style, which means that it should also replace existing `if let Some(x) = y` to `if y is Some(x)`. Regardless of whether the `is` operator is a more reasonable expression, we cannot remove the existing `if let` style, so extending it is still meaningful.

**Comment by CEbbinghaus** [comment]
After reading rust-lang/rfcs#3573 I actually disagree with @TennyZhuang. The two features stand in opposition to one another as they both offer the same functionality. However the `is` operator is significantly more flexible and consistent due to it returning a boolean which lets it fit anywhere a boolean can. Currently the only thing that cannot be achieved any other way is chaining pattern matching which this RFC addresses. As such if this RFC makes it to stable then there is 0 functional reason to add `is`, only ergonomic reasons (which already has a vocal opposition).

**Comment by Nilstrieb** [comment]
Please remember that tracking issues are *not* for discussion, but for updates, especially large ones like this one. Please open a new issue/Zulip/internals.rust-lang.org thread (that can be linked from here) to discuss the feature, do that people subscribed to the issue don't get spammed and the issue history remains readable.

**Comment by traviscross** [comment]
We talked about this in the lang planning meeting today.  As far as we know, the blockers and semi-blockers to this have been resolved.  E.g.:

- https://github.com/rust-lang/rust/issues/123742
- https://github.com/rust-lang/rust/issues/124085

At this point, we would encourage someone to write up a comprehensive stabilization report and include that in a PR proposing `let-chains` for stabilization.  We're looking forward to considering that proposal.

**Comment by est31** [comment]
This is really wonderful news, @traviscross! I've went over the [original stabilization PR](https://github.com/rust-lang/rust/issues/131154) (now reverted) and related issues to refresh my memory. I've found two possible stabilization blockers that I have opened zulip threads about.

* [syntactic rejection is maybe not complete?](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/new.20shot.20at.20let.20chains.20stabilization)
* [formatting stabilization](https://rust-lang.zulipchat.com/#narrow/channel/346005-t-style/topic/let.20chains.20stabilization.20and.20formatting): from [how I read this document](https://github.com/rust-lang/style-team/blob/main/nightly-style-procedure.md), nightly syntax formatting needs to be stabilized and formalized. We have formatting present in rustfmt but is it also formalized?
* I don't see anything that should be a stabilization blocker with the [`F-let_chains`](https://github.com/rust-lang/rust/labels/F-let_chains) label at this point

As there is many people subscribed, please try to keep replies in the linked zulip threads.

**Comment by est31** [comment]
Stabilization report: https://github.com/rust-lang/rust/issues/132833

**Comment by xXA2005** [comment]
fix pls thx

`let` expressions in this position are unstable
see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more informationrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[0]?0#file:///home/ab5/codes/procap/src/storage.rs)

**Comment by RalfJung** [comment]
There is still an "unstable" feature in the Rust compiler pointing to this issue as its tracking issue:

https://github.com/rust-lang/rust/blob/eb7245a864613f290c5668f4f4e8376edd065dc6/compiler/rustc_feature/src/unstable.rs#L555-L556

What is up with that? We shouldn't have unstable features point to closed issues.

**Comment by thaliaarchi** [comment]
@RalfJung This unstable feature is a special case, retained for now to ease the transition for editions before 2024. See https://github.com/rust-lang/rust/pull/132833#issuecomment-2853828018 and https://github.com/rust-lang/rust/issues/140722.

**Comment by RalfJung** [comment]
Okay... but procedurally, it doesn't make sense to have an active feature pointing at a closed issue. (Ideally we'd auto-detect this.)

Either the feature should then point at https://github.com/rust-lang/rust/issues/140722, or this issue here should be reopened and the issue text updated to reflect the current status. I do not have a preference either way.

**Comment by inquisitivecrystal** [comment]
#143214 would close this out completely. If that can be reviewed and merged soonish, it would be the ideal resolution. It might be worth leaving this in a procedurally incorrect state for a little while to let people finish and merge that PR.

**Comment by bananaturtlesandwich** [comment]
is this feature also planned for let else statements?

**Comment by camsteffen** [comment]
> Okay... but procedurally, it doesn't make sense to have an active feature pointing at a closed issue.

It also potentially confuses users because error output points to the tracking issue.

**Comment by not-matthias** [comment]
To anyone else getting redirected to this issue even though you are already using Rust 1.88: Make sure you use the 2024 edition! 

(shouldn't this be an error or documented somewhere?)

**Comment by inquisitivecrystal** [comment]
> To anyone else getting redirected to this issue even though you are already using Rust 1.88: Make sure you use the 2024 edition!
> 
> (shouldn't this be an error or documented somewhere?)

That's also part of #143214!

**Comment by est31** [comment]
@not-matthias as @inquisitivecrystal , it's being made into a proper error in #143214, tracked by #140722, and marked as stabilized in that same PR.

The stabilization of edition specific things is unusual. There is a number of nightly users I didn't want to break, most importantly the compiler itself and associated components like miri. Had we wanted to mark it as stabilized right away, one would have needed to wait for all of those components to be on edition 2024, or done the edition update in my PR. All of this right after the release of edition 2024.

It is even more unusual for the compiler and associated components to make such heavy use of a feature. The one other edition specific feature that I know about, async/await syntax, was almost not used in the compiler for comparison. This is neither good or bad. I suppose that let chains are just more useful for the compiler use case than async await, but it means that more components needed to be "in sync".

The non-compiler nightly users of let chains would also have needed to immediately upgrade their edition or been unable to upgrade their nightly: this way they had some leeway as well. Balancing this with temporary sub-optimal error messages, I preferred not having the perfect error story right after stabilization, and until yesterday, nobody has really complained either.

PR #143214 is not merged yet, but will will likely ship in 1.90.0, so in 2 months. Until then the error situation will be a bit suboptimal. It's normal and to be expected that stuff is not 100% smooth right after stabilization of a new feature.

**Comment by camsteffen** [comment]
Could we have separated the addition of the 2024 requirement from the feature stabilization?

**Comment by est31** [comment]
Yeah we could probably have gated only if the feature is already active, and if it's not active, not gate but immediately error with the custom error you add in https://github.com/rust-lang/rust/pull/143214.

**Comment by est31** [comment]
Added a text to the top of the issue to make things more clear to users, so that we don't get the same already addressed complaint over and over. It's probably only needed temporarily.

**Comment by RalfJung** [comment]
This is not a support forum. Please try the Rust community discord or https://users.rust-lang.org/ .

**Comment by estebank** [comment]
@RalfJung To be fair, the diagnostic *did* tell them to come here. :)

I suspect it might be a good idea at some point to implement a mechanism for older toolchains to learn about when features were stabilized so they can provide more appropriate instructions.
